### ISSUE TITLE
Adding Mouse Sensitivity Setting in Moonlight?

### ISSUE THREAD

**Issue by somerandomperson123** [moonlight-stream/moonlight-qt#128]
I had a previous post regarding this - https://github.com/moonlight-stream/moonlight-qt/issues/116

but that seemed to go off topic to discussing mouse acceleration/mouse input.

What I'm wondering about is the possibility of Moonlight itself having a sensitivity option that modifies mouse sensitivity? 

Perhaps this is only an issue that affects me. My problem is that I run Moonlight on PC with a lower DPI mouse and a lower resolution monitor compared to my home PC (host).

**Comment by Zibbibbo** [comment]
Same problem for me. For now i've managed the problem using a mouse with switchable DPI on the client pc, so, when i take the host control, I switch to higher DPI and voilàt

**Comment by Ravstin** [comment]
Yes, that would be really useful, hopefully that option would be added

**Comment by installgentoo** [comment]
I would appreciate moonlight just not changing my mouse settings every time.

**Comment by cgutman** [comment]
@installgentoo It's not something we can control. See #383

**Comment by SalvatoryMalamar** [comment]
@cgutman I understand that enhanced mouse pointer precision changes by Geforce Experience can't be prevented by moonlight. Instead, after the client connects Is it possible to change the Windows host's enhanced pointer precision in order to mitigate issue #116? In addition, mouse speed could be set at this time to accommodate @somerandomperson123's issue with differences in cursor input between client and host.

For example, in moonlight settings there could be a host mouse sensitivity slider and enhanced pointer precision checkbox. If the option to use such settings is enabled then moonlight sets those settings on the PC after the client connects. An additional option may be needed to return the host PC's mouse settings to a different setting after the client disconnects.

In my research I found [this stackoverflow exchange](https://stackoverflow.com/questions/16813653/mouse-speed-not-changing-by-using-spi-setmousespeed) discussing how to change mouse speed in C++, and [this post](https://stackoverflow.com/questions/3902477/how-to-configure-mouse-enhance-pointer-precision-programmatically) discussing how to change the enhanced pointer precision setting in C++.

All the above is about such a feature is possible, but not if it's desirable. Would it be desirable to incorporate such a feature? There appears to be only two issues regarding mouse speed #33 #125, and four regarding enhanced mouse pointer precision #480 #383 #314 #258. 

If it is desirable to add such a feature, how could the UI be setup so the feature is as simple and intuitive as possible? And, what would happen if the host closes moonlight, or moonlight crashes, before the client disconnects? Would the mouse settings remain at the modified ones? Is that acceptable?

**Comment by installgentoo** [comment]
@SalvatoryMalamar just add a "set host mouse settings" checkbox and then user's responsible for resetting them back.

User can then also just set moonlight mouse settings to whatever he has on his desktop in the first place, and you won't have to reset settings every time.

**Comment by StuartMinkus** [comment]
@SalvatoryMalamar Signed up to say that feature would help me big time! Streaming from my 1920x1080 PC to a Shield TV 65"4k, my DS4 controller with the mouse emulation on is completely out of whack! High sensitivity and acceleration make it nearly impossible to change the host settings without using the host settings manually.

**Comment by Tsubajashi** [comment]
I would also love that function. I mean, I can fix it with a bat to disable it, but that shouldn't be an issue to begin with.

**Comment by mrwiwi** [comment]
Check that fix :

Even with the box ticked, no acceleration !

http://donewmouseaccel.blogspot.com/2010/03/markc-windows-7-mouse-acceleration-fix.html?m=1

I never use it so I don't care, it's a workaround for old games that enable mouse acceleration...

**Comment by predprey2** [comment]
You can edit "C:\Program Files\NVIDIA Corporation\NvStreamSrv\nvstreamer.exe" directly to disable it calling SystemParametersInfoA to set mouse acceleration.

Just replace the bytes "FF 15 7F 62 0A 00" with "B8 01 00 00 00 90" using a hex editor like HxD.

Tested on GeForce Experience 3.24.0.123, nvstreamer.exe file version 7.1.3048.4816.

Since these are relative offsets, they may not change with versions, but the steps to reproduce are to just analyze the executable and search for all instances of the referenced function. On this version, it was the third last instance, then just replace it with the same bytes for it to spoof the return success flag.

#513

**Comment by sivan-koren** [comment]
I found this PowerShell snippet somewhere on Stack Overflow. Running this in PowerShell will disable Enhanced Pointer Precision. This is not permanent, it's equivalent to unchecking the "Enhance Pointer Precision" box in main.cpl.

I used PS2EXE module in PowerShell 5 to make an exe out of it. I've confirmed that the exe does work when executed.

Just need some way to trigger it on connection. Task Scheduler didn't seem to work for me, because I couldn't find any event to link the trigger to in Event Viewer. Maybe someone else can get it to work. Please post back if so.

> $code=@'
[DllImport("user32.dll", EntryPoint = "SystemParametersInfo")]
 public static extern bool SystemParametersInfo(uint uiAction, uint uiParam, int[] pvParam, uint fWinIni);
'@
Add-Type $code -name Win32 -NameSpace System
[System.Win32]::SystemParametersInfo(4,0,0,2)

**Comment by samtowers** [comment]
> Strange. I am also unable to find that pattern in the nvstreamer.exe. I just tried this with Hextor 0.13, on nvstreamer.exe 7.1.3110.6651, checking the HEX box in the find dialog, and got "search pattern not found." As a test to confirm that Hextor is finding other patterns, I searched for 01 FA E6 0C 00 8F 84 0A and did find a result.

I found it at address 25D650 (on the left):

![image](https://user-images.githubusercontent.com/8900168/168136669-8e2fca5b-2eac-4d85-a4e3-f8346036962e.png)

The highlighted hex shows 'B8 01 00 00 00 90' as I've already replaced it. Previously this was  'FF 15 FF 60 0A 00'.

**Comment by sivan-koren** [comment]
I find a very similar string at that address, but the 3rd byte is 4F not FF.

The solution may be to search and change FF 15 4F 60 0A 00 to B8 01 00 00 00 90.

Not tested. Screenshot below for the address on my nvstreamer.exe.
![Screenshot 2022-05-12 115903](https://user-images.githubusercontent.com/13458661/168139710-85417613-1552-49ac-aaf3-1b3ba903e5d4.png)

**Comment by predprey2** [comment]
> UPDATE: This doesn't work!
> 
> I find a very similar string at that address, but the 3rd byte is 4F not FF.
> 
> The solution may be to search and change FF 15 4F 60 0A 00 to B8 01 00 00 00 90.
> 
> Not tested. Screenshot below for the address on my nvstreamer.exe. ![Screenshot 2022-05-12 115903](https://user-images.githubusercontent.com/13458661/168139710-85417613-1552-49ac-aaf3-1b3ba903e5d4.png)

GeForce Experience 3.26.0.131, nvstreamer.exe 7.1.3164.8965.
Replace the bytes "FF 15 1A DE 1A 00" with "B8 01 00 00 00 90"

**Comment by 29988122** [comment]
I think this patching method is very useful, rather than waiting for nvidia to add this option and fix. Thanks!

I wonder, is there any way to ida pro/make a patcher to find relevant code segments, and replace/patch the bytes to MOV EAX NOP directly?

**Comment by predprey2** [comment]
> I think this patching method is very useful, rather than waiting for nvidia to add this option and fix. Thanks!
> 
> I wonder, is there any way to ida pro/make a patcher to find relevant code segments, and replace/patch the bytes to MOV EAX NOP directly? (Assuming there's anther new version which changes it)

Probably can look into [Zydis disassembler](https://github.com/zyantific/zydis) to find specific structure offsets at runtime to target the subroutines that calls the mouse sensitivity WinAPI functions. This should be more stable than machine code. The injection method would be by hijacking one of the commonly WinAPI loaded libraries, placing the DLL in the same folder as nvstreamer.exe or nvcontainer.exe. The DLL would then proxy the original functions in that library while searching for the code signature and injecting the NOPs.

**Comment by 29988122** [comment]
Thank you, predprey2, for the very extensive reply!
I think your thoughts of suggesting me to tackle via "different layers than machine code" is very inspiring to me.
It's a very learning-worthy instruction, and I can't thank you more about it! : DDD

On the other hand, it's sad to see nvidia soon dropping support to gamestream. It's a very useful way to wireless stream desktop to TVs with Android, and with higher quality than steamlink.

Hope that we'll have a brighter future to Sunshine, and kudos to the moonlight team for it - it's not easy feat.

**Comment by Autliety** [comment]
> GeForce Experience 3.26.0.131, nvstreamer.exe 7.1.3164.8965. Replace the bytes "FF 15 1A DE 1A 00" with "B8 01 00 00 00 90"

Not avalible in GeForce Experience 3.27.0.112, nvstreamer.exe 7.1.3219.3813

**Comment by sivan-koren** [comment]
I've disabled updates by blocking them in the hosts file. This seems like a
reasonable solution where updates will be disabling the function of
Moonlight altogether soon anyway.
Until Sunshine is improved...
